snuba/query/parser/conditions.py

```diff
@@ -0,0 +1,225 @@
+from typing import Any, Callable, Optional, OrderedDict, Sequence, TypeVar
+
+from snuba.datasets.dataset import Dataset
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    Expression,
+    Lambda,
+    Literal,
+    FunctionCall,
+)
+from snuba.query.conditions import (
+    binary_condition,
+    BooleanFunctions,
+    OPERATOR_TO_FUNCTION,
+)
+from snuba.query.parser.functions import parse_function_to_expr
+from snuba.query.schema import POSITIVE_OPERATORS
+from snuba.util import is_condition, is_function, QUOTED_LITERAL_RE
+
+
+TExpression = TypeVar("TExpression")
+
+
+class InvalidConditionException(Exception):
+    pass
+
+
+def parse_conditions(
```

Comment:
This is essentially a cut/paste of the current algorithm, but it is parametric. So it preserves all the corner cases.